src/lib.rs

```diff
@@ -2,269 +2,157 @@
 //
 // SPDX-License-Identifier: MIT OR Apache-2.0
 
-/*! The library provide tools for handling compressed and archive files
-
-# Examples
-```no_run
-use compress_tools;
-
-let dir = tempfile::tempdir().unwrap();
-compress_tools::uncompress("tests/fixtures/tree.tar.gz", dir.path(), compress_tools::Kind::TarGZip).unwrap();
-```
-*/
-
-mod uncompress;
-
-pub use uncompress::{Error, Result};
-
-use std::path::Path;
-
-/// Type of compressed file or archive
-#[derive(Copy, Clone, Debug)]
-pub enum Kind {
-    TarGZip,
-    TarBZip2,
-    TarXz,
-    TarLZMA,
-    TarLZip,
-    Tar,
-    Zip,
-
-    GZip,
-    BZip2,
-    Xz,
-    LZip,
-    LZMA,
-}
-
-/// Uncompress a archive of known [Kind](Kind) pointed by source to target
-/// location for file
-pub fn uncompress<P1, P2>(source: P1, target: P2, kind: Kind) -> Result<()>
-where
-    P1: AsRef<Path>,
-    P2: AsRef<Path>,
-{
-    let source = uncompress::Uncompress::new(source.as_ref());
-    let target = target.as_ref();
-    match kind {
-        // Single files
-        Kind::BZip2 => source.bz2().file(target),
-        Kind::GZip => source.gz().file(target),
-        Kind::LZMA => source.lzma().file(target),
-        Kind::LZip => source.lzip().file(target),
-        Kind::Xz => source.xz().file(target),
-
-        // Archive files
-        Kind::Tar => source.tar(target),
-        Kind::TarBZip2 => source.bz2().tar(target),
-        Kind::TarGZip => source.gz().tar(target),
-        Kind::TarLZMA => source.lzma().tar(target),
-        Kind::TarLZip => source.lzip().tar(target),
-        Kind::TarXz => source.xz().tar(target),
-        Kind::Zip => source.unzip(target),
+#![allow(non_upper_case_globals)]
+#![allow(non_camel_case_types)]
+#![allow(non_snake_case)]
+
+pub mod error;
+pub mod ffi;
+pub mod util;
+
+use error as compress;
+use ffi::*;
+use std::{
+    ffi::{CStr, CString},
+    os::raw::c_void,
+};
+
+// reads the input buffer and extract to another buffer.
+pub fn buffered_uncompress(source: &[u8], target: &mut [u8]) -> compress::Result<()> {
+    let mut entry: *mut archive_entry = std::ptr::null_mut();
+
+    unsafe {
+        let (ar, aw) = util::init_archive_rw()?;
```

Comment:
You could rewrite this using a lambda. I did something similar for `nix` at:

https://github.com/nix-rust/nix/pull/1139/commits/b14415a541c8de9569a720244b6b28624695d7c4#diff-93e5e3ed2568f34ed9de08d04a0443c2R2601-R2634